GVFS/GVFS.Platform.Windows/WindowsFileSystemVirtualizer.cs

```diff
@@ -1029,7 +1035,7 @@ private HResult QueryFileNameHandler(string virtualPath)
                                     throw new GetFileStreamException("IOException while copying to unmanaged buffer: " + e.Message, (HResult)HResultExtensions.HResultFromNtStatus.FileNotAvailable);
                                 }
 
-                                long writeOffset = length - remainingData;
+                                long writeOffset = blobLength - remainingData;

```

Comment:
> Do we not need to worry about the writeOffset being past the length that was passed in?

That scenario is OK.  Christian explained that the provider needs to provide at least `offset` to `offset+length` data when it receives the file stream callback.  It's OK to provide more data than the caller requested.

> What is the purpose the length field that gets passed in now?

In the current version of ProjFS the length field will always be the length of the file *except* when the file > 4GB in size.  Once the file crosses that threshold ProjFS will split the file stream callbacks into chunks.  

In future versions of ProjFS that support partial recall, providers will have the option (VFS4G won't choose this option) of getting callbacks where the offset and length are based on what the caller is wanting to read.

An interesting side note:  VFS4G had trouble dealing with 4GB+ sized files (outside of this issue) when I was testing.  I'm wondering if there is a limitation in libgit2 that prevents it from dealing with 4GB files properly.

> Doesn't it seem like we can just get rid of all that code?

Currently we're only logging some extra info with the length, is that the code you're thinking about (or did you mean removing it from the API completely)?

Reply:
> An interesting side note: VFS4G had trouble dealing with 4GB+ sized files (outside of this issue) when I was testing. I'm wondering if there is a limitation in libgit2 that prevents it from dealing with 4GB files properly.

I think _Git for Windows_ has problems with 4GB files because of `unsigned long`s all over the codebase, and Windows compiles them to 32-bit uints. [LibGit2 also has `unsigned long` everywhere](https://github.com/libgit2/libgit2/search?utf8=%E2%9C%93&q=unsigned+long&type=). I think neither will be fast to fix this problem, unfortunately.

Reply:
>Currently we're only logging some extra info with the length, is that the code you're thinking about (or did you mean removing it from the API completely)?

I was wondering but it sounds like this might be used for a future version of ProjFS and if the logging is useful for debugging we can keep it but we aren't using it for anything in the code so not sure it is giving any value..

Reply:
>>   I think neither will be fast to fix this problem, unfortunately.

> Thanks for the details! I don't think there's any pressing need to fix this (no one is asking for it). It does mean though that these changes are more about preparing for partial recall than they are for supporting 4GB+ files (since we have other issues with files that big).

My guess is that they need this feature for One Drive. It does present an interesting issue, though: now that we are going cross-platform a Mac user could push a 5GB file that a Windows user could have trouble reading. I think we have enough server-side checks for large pushes, though.